vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenFragment.kt

```diff
@@ -59,7 +59,7 @@ class LockScreenFragment : VectorBaseFragment<FragmentLockScreenBinding>() {
             if (state.lockScreenConfiguration.mode == LockScreenMode.CREATE) return@withState
 
             viewLifecycleOwner.lifecycleScope.launchWhenResumed {
-                if (state.isBiometricKeyInvalidated) {
+                if (state.canUseBiometricAuth && state.isBiometricKeyInvalidated) {
                     lockScreenListener?.onBiometricKeyInvalidated()
```

Comment:
out of interest, what happens in this `onBiometricKeyInvalidated` flow?

Reply:
There are 2 points where the `state. isBiometricKeyInvalidated ` can change and trigger this code:

1. `LockScreenViewModel.showBiometricPrompt`, this will call `BiometricHelper.authenticate` that will fail before showing any UI if the system key is not valid and instead it:
* Catches any exceptions.
* Posts a `LockScreenViewEvent.AuthError` to handle it in the UI.
* If the error was an instance of `KeyPermanentlyInvalidatedException`, it will call `removeBiometricAuthentication()` which will disable the 'biometric enabled' option, delete the broken system key and update the current state to display an alert saying that the key has just been invalidated.

2. `LockScreenViewModel.updateStateWithBiometricInfo` is automatically called when the  `LockScreenConfiguration` changes:
* It calls `biometricHelper.isSystemKeyValid`.
* This will reach `KeyStoreCrypto.hasValidKey()`, which tries to retrieve the current system key and either return true or catch both `KeyPermanentlyInvalidatedException` and `UserNotAuthenticatedException`, which are thrown when the internal `Cipher.init` fails, and return false.
* `state.isBiometricKeyInvalidated` will be updated with this returned value and used to display the same alert as in the case above.

Reply:
thanks for explaining :+1: